File: C++/TaskManager/TaskManager.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <ctime>
#include <algorithm>
#include <iomanip>
#include <windows.h>
#include <map>

#define RESET   "\033[0m"
#define RED     "\033[31m"
#define GREEN   "\033[32m"
#define YELLOW  "\033[33m"
#define BLUE    "\033[34m"
#define MAGENTA "\033[35m"
#define CYAN    "\033[36m"

class Task {
public:
    std::string description;
    std::string deadline;
    int priority; 
    bool completed;
    std::string category;
    std::string notes;
    time_t created_at;

    Task(const std::string& desc, const std::string& dl, int prio, const std::string& cat = "General")
        : description(desc), deadline(dl), priority(prio), completed(false), 
          category(cat), notes(""), created_at(time(nullptr)) {}
};

class TaskManager {
private:
    std::vector<Task> tasks;
    const std::string filename = "tasks.txt";
    std::map<std::string, int> categoryStats;

    void saveTasks() {
        std::ofstream file(filename);
        if (file.is_open()) {
            for (const auto& task : tasks) {
                file << task.description << "\n"
                     << task.deadline << "\n"
                     << task.priority << "\n"
                     << task.completed << "\n"
                     << task.category << "\n"
                     << task.notes << "\n"
                     << task.created_at << "\n";
            }
        }
    }

    void loadTasks() {
        std::ifstream file(filename);
        if (file.is_open()) {
            tasks.clear();
            std::string desc, deadline, category, notes;
            int prio;
            bool completed;
            time_t created_at;
            while (std::getline(file, desc) &&
                   std::getline(file, deadline) &&
                   file >> prio >> completed &&
                   file.ignore() &&
                   std::getline(file, category) &&
                   std::getline(file, notes) &&
                   file >> created_at) {
                tasks.emplace_back(desc, deadline, prio, category);
                tasks.back().completed = completed;
                tasks.back().notes = notes;
                tasks.back().created_at = created_at;
                file.ignore();
            }
        }
        updateCategoryStats();
    }

    void updateCategoryStats() {
        categoryStats.clear();
        for (const auto& task : tasks) {
            categoryStats[task.category]++;
        }
    }

    std::string getCurrentDate() {
        time_t now = time(nullptr);
        tm* localTime = localtime(&now);
        char buffer[11];
        strftime(buffer, sizeof(buffer), "%Y-%m-%d", localTime);
        return std::string(buffer);
    }

    std::string formatDate(const std::string& date) {
        if (date == "today") return getCurrentDate();
        return date;
    }

    void printTask(const Task& task, int index) {
        std::string priorityColor;
        switch (task.priority) {
            case 1: priorityColor = RED; break;
            case 2: priorityColor = YELLOW; break;
            case 3: priorityColor = GREEN; break;
            default: priorityColor = RESET;
        }

        std::cout << std::setw(3) << index + 1 << ". "
                  << (task.completed ? GREEN "[X]" RESET : RED "[ ]" RESET) << " "
                  << priorityColor << "Priority: " << std::string(task.priority, '!') << RESET << " "
                  << BLUE << "Category: " << task.category << RESET << "\n"
                  << "    Deadline: " << (isOverdue(task) ? RED : CYAN) << task.deadline << RESET << "\n"
                  << "    " << task.description << "\n";
        if (!task.notes.empty()) {
            std::cout << "    Notes: " << MAGENTA << task.notes << RESET << "\n";
        }
    }

    bool isOverdue(const Task& task) {
        if (task.completed) return false;
        time_t now = time(nullptr);
        tm* localTime = localtime(&now);
        char currentDate[11];
        strftime(currentDate, sizeof(currentDate), "%Y-%m-%d", localTime);
        return task.deadline < currentDate;
    }

    void printStatistics() {
        std::cout << "\n=== Task Statistics ===\n";
        std::cout << "Total tasks: " << tasks.size() << "\n";
        
        int completed = std::count_if(tasks.begin(), tasks.end(), 
            [](const Task& t) { return t.completed; });
        std::cout << "Completed tasks: " << completed << "\n";
        std::cout << "Completion rate: " 
                  << std::fixed << std::setprecision(1)
                  << (tasks.empty() ? 0 : (completed * 100.0 / tasks.size()))
                  << "%\n";

        std::cout << "\nTasks by category:\n";
        for (const auto& stat : categoryStats) {
            std::cout << "  " << stat.first << ": " << stat.second << "\n";
        }

        int overdue = std::count_if(tasks.begin(), tasks.end(),
            [this](const Task& t) { return isOverdue(t); });
        std::cout << "\nOverdue tasks: " << overdue << "\n";
    }

public:
    TaskManager() {
        loadTasks();
    }

    ~TaskManager() {
        saveTasks();
    }

    void addTask() {
        std::string desc, deadline, category, notes;
        int priority;

        std::cout << "Enter task description: ";
        std::cin.ignore();
        std::getline(std::cin, desc);

        std::cout << "Enter deadline (YYYY-MM-DD) or 'today': ";
        std::cin >> deadline;
        deadline = formatDate(deadline);

        std::cout << "Enter category (default: General): ";
        std::cin.ignore();
        std::getline(std::cin, category);
        if (category.empty()) category = "General";

        std::cout << "Enter priority (1-3, 1 is highest): ";
        std::cin >> priority;
        priority = std::max(1, std::min(3, priority));

        std::cout << "Enter notes (optional): ";
        std::cin.ignore();
        std::getline(std::cin, notes);

        tasks.emplace_back(desc, deadline, priority, category);
        tasks.back().notes = notes;
        updateCategoryStats();
        std::cout << GREEN << "Task added successfully!" << RESET << "\n";
    }

    void listTasks(const std::string& filter = "", const std::string& sortBy = "priority") {
        if (tasks.empty()) {
            std::cout << "No tasks found.\n";
            return;
        }

        std::vector<Task> filteredTasks = tasks;
        

        if (!filter.empty()) {
            filteredTasks.erase(
                std::remove_if(filteredTasks.begin(), filteredTasks.end(),
                    [&filter](const Task& t) {
                        return t.description.find(filter) == std::string::npos &&
                               t.category.find(filter) == std::string::npos &&
                               t.notes.find(filter) == std::string::npos;
                    }),
                filteredTasks.end()
            );
        }

        if (sortBy == "priority") {
            std::sort(filteredTasks.begin(), filteredTasks.end(),
                [](const Task& a, const Task& b) {
                    if (a.priority != b.priority)
                        return a.priority < b.priority;
                    return a.deadline < b.deadline;
                });
        } else if (sortBy == "deadline") {
            std::sort(filteredTasks.begin(), filteredTasks.end(),
                [](const Task& a, const Task& b) {
                    return a.deadline < b.deadline;
                });
        } else if (sortBy == "category") {
            std::sort(filteredTasks.begin(), filteredTasks.end(),
                [](const Task& a, const Task& b) {
                    return a.category < b.category;
                });
        }

        std::cout << "\n=== Your Tasks ===\n";
        for (size_t i = 0; i < filteredTasks.size(); ++i) {
            printTask(filteredTasks[i], i);
        }
    }

    void editTask() {
        if (tasks.empty()) {
            std::cout << "No tasks to edit.\n";
            return;
        }

        listTasks();
        std::cout << "Enter task number to edit: ";
        int index;
        std::cin >> index;
        index--;

        if (index >= 0 && index < static_cast<int>(tasks.size())) {
            std::string desc, deadline, category, notes;
            int priority;

            std::cout << "Enter new description (press Enter to keep current): ";
            std::cin.ignore();
            std::getline(std::cin, desc);
            if (!desc.empty()) tasks[index].description = desc;

            std::cout << "Enter new deadline (YYYY-MM-DD or 'today', press Enter to keep current): ";
            std::getline(std::cin, deadline);
            if (!deadline.empty()) tasks[index].deadline = formatDate(deadline);

            std::cout << "Enter new category (press Enter to keep current): ";
            std::getline(std::cin, category);
            if (!category.empty()) tasks[index].category = category;

            std::cout << "Enter new priority (1-3, press Enter to keep current): ";
            std::string prioStr;
            std::getline(std::cin, prioStr);
            if (!prioStr.empty()) {
                priority = std::stoi(prioStr);
                tasks[index].priority = std::max(1, std::min(3, priority));
            }

            std::cout << "Enter new notes (press Enter to keep current): ";
            std::getline(std::cin, notes);
            if (!notes.empty()) tasks[index].notes = notes;

            updateCategoryStats();
            std::cout << GREEN << "Task updated successfully!" << RESET << "\n";
        } else {
            std::cout << RED << "Invalid task number." << RESET << "\n";
        }
    }

    void markCompleted() {
        if (tasks.empty()) {
            std::cout << "No tasks to mark.\n";
            return;
        }

        listTasks();
        std::cout << "Enter task number to mark as completed: ";
        int index;
        std::cin >> index;
        index--;

        if (index >= 0 && index < static_cast<int>(tasks.size())) {
            tasks[index].completed = true;
            std::cout << GREEN << "Task marked as completed!" << RESET << "\n";
        } else {
            std::cout << RED << "Invalid task number." << RESET << "\n";
        }
    }

    void deleteTask() {
        if (tasks.empty()) {
            std::cout << "No tasks to delete.\n";
            return;
        }

        listTasks();
        std::cout << "Enter task number to delete: ";
        int index;
        std::cin >> index;
        index--;

        if (index >= 0 && index < static_cast<int>(tasks.size())) {
            tasks.erase(tasks.begin() + index);
            updateCategoryStats();
            std::cout << GREEN << "Task deleted successfully!" << RESET << "\n";
        } else {
            std::cout << RED << "Invalid task number." << RESET << "\n";
        }
    }

    void searchTasks() {
        std::string query;
        std::cout << "Enter search query: ";
        std::cin.ignore();
        std::getline(std::cin, query);
        listTasks(query);
    }

    void showMenu() {
        std::cout << "\n=== Task Manager ===\n"
                  << "1. Add new task\n"
                  << "2. List all tasks\n"
                  << "3. Search tasks\n"
                  << "4. Edit task\n"
                  << "5. Mark task as completed\n"
                  << "6. Delete task\n"
                  << "7. Show statistics\n"
                  << "8. Exit\n"
                  << "Choose an option: ";
    }
};

int main() {
    
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
    DWORD mode = 0;
    GetConsoleMode(hConsole, &mode);
    SetConsoleMode(hConsole, mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);

    TaskManager manager;
    int choice;

    while (true) {
        manager.showMenu();
        std::cin >> choice;

        switch (choice) {
            case 1:
                manager.addTask();
                break;
            case 2:
                manager.listTasks();
                break;
            case 3:
                manager.searchTasks();
                break;
            case 4:
                manager.editTask();
                break;
            case 5:
                manager.markCompleted();
                break;
            case 6:
                manager.deleteTask();
                break;
            case 7:
                manager.printStatistics();
                break;
            case 8:
                return 0;
            default:
                std::cout << RED << "Invalid option. Please try again." << RESET << "\n";
        }
    }

    return 0;
} 

File: Python/TaskManager/TaskManager.py
import json
import os
from datetime import datetime
from typing import List, Dict, Optional

class Task:
    def __init__(self, title: str, description: str = "", priority: int = 1, 
                 deadline: Optional[str] = None, completed: bool = False):
        self.title = title
        self.description = description
        self.priority = priority
        self.deadline = deadline
        self.completed = completed
        self.created_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        self.completed_at = None

    def to_dict(self) -> Dict:
        return {
            "title": self.title,
            "description": self.description,
            "priority": self.priority,
            "deadline": self.deadline,
            "completed": self.completed,
            "created_at": self.created_at,
            "completed_at": self.completed_at
        }

    @classmethod
    def from_dict(cls, data: Dict) -> 'Task':
        task = cls(
            title=data["title"],
            description=data["description"],
            priority=data["priority"],
            deadline=data["deadline"],
            completed=data["completed"]
        )
        task.created_at = data["created_at"]
        task.completed_at = data["completed_at"]
        return task

    def __str__(self) -> str:
        status = "✓" if self.completed else "✗"
        priority_str = "★" * self.priority
        deadline_str = f" | Дедлайн: {self.deadline}" if self.deadline else ""
        return f"{status} {priority_str} {self.title}{deadline_str}"

class TaskManager:
    def __init__(self, filename: str = "tasks.json"):
        self.filename = filename
        self.tasks: List[Task] = []
        self.load_tasks()

    def load_tasks(self) -> None:
        if os.path.exists(self.filename):
            try:
                with open(self.filename, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.tasks = [Task.from_dict(task_data) for task_data in data]
            except Exception as e:
                print(f"Ошибка при загрузке задач: {e}")
                self.tasks = []

    def save_tasks(self) -> None:
        try:
            with open(self.filename, 'w', encoding='utf-8') as f:
                json.dump([task.to_dict() for task in self.tasks], f, 
                         ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"Ошибка при сохранении задач: {e}")

    def add_task(self) -> None:
        print("\n=== Добавление новой задачи ===")
        title = input("Введите название задачи: ").strip()
        if not title:
            print("Название задачи не может быть пустым!")
            return

        description = input("Введите описание задачи (необязательно): ").strip()
        
        while True:
            try:
                priority = int(input("Введите приоритет (1-5): ").strip())
                if 1 <= priority <= 5:
                    break
                print("Приоритет должен быть от 1 до 5!")
            except ValueError:
                print("Введите число!")

        deadline = input("Введите дедлайн (ГГГГ-ММ-ДД или пусто): ").strip()
        if deadline:
            try:
                datetime.strptime(deadline, "%Y-%m-%d")
            except ValueError:
                print("Неверный формат даты! Используйте ГГГГ-ММ-ДД")
                return

        task = Task(title, description, priority, deadline if deadline else None)
        self.tasks.append(task)
        self.save_tasks()
        print("Задача успешно добавлена!")

    def edit_task(self) -> None:
        if not self.tasks:
            print("Список задач пуст!")
            return

        self.show_tasks()
        try:
            index = int(input("\nВведите номер задачи для редактирования: ")) - 1
            if 0 <= index < len(self.tasks):
                task = self.tasks[index]
                print("\n=== Редактирование задачи ===")
                print("Оставьте поле пустым, чтобы не изменять значение")
                
                title = input(f"Название [{task.title}]: ").strip()
                if title:
                    task.title = title
                
                description = input(f"Описание [{task.description}]: ").strip()
                if description:
                    task.description = description
                
                priority = input(f"Приоритет (1-5) [{task.priority}]: ").strip()
                if priority:
                    try:
                        priority = int(priority)
                        if 1 <= priority <= 5:
                            task.priority = priority
                        else:
                            print("Приоритет должен быть от 1 до 5!")
                    except ValueError:
                        print("Введите число!")
                
                deadline = input(f"Дедлайн (ГГГГ-ММ-ДД) [{task.deadline}]: ").strip()
                if deadline:
                    try:
                        datetime.strptime(deadline, "%Y-%m-%d")
                        task.deadline = deadline
                    except ValueError:
                        print("Неверный формат даты! Используйте ГГГГ-ММ-ДД")
                
                self.save_tasks()
                print("Задача успешно отредактирована!")
            else:
                print("Неверный номер задачи!")
        except ValueError:
            print("Введите число!")

    def delete_task(self) -> None:
        if not self.tasks:
            print("Список задач пуст!")
            return

        self.show_tasks()
        try:
            index = int(input("\nВведите номер задачи для удаления: ")) - 1
            if 0 <= index < len(self.tasks):
                task = self.tasks.pop(index)
                self.save_tasks()
                print(f"Задача '{task.title}' удалена!")
            else:
                print("Неверный номер задачи!")
        except ValueError:
            print("Введите число!")

    def toggle_task(self) -> None:
        if not self.tasks:
            print("Список задач пуст!")
            return

        self.show_tasks()
        try:
            index = int(input("\nВведите номер задачи для изменения статуса: ")) - 1
            if 0 <= index < len(self.tasks):
                task = self.tasks[index]
                task.completed = not task.completed
                task.completed_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S") if task.completed else None
                self.save_tasks()
                status = "выполнена" if task.completed else "не выполнена"
                print(f"Задача '{task.title}' помечена как {status}!")
            else:
                print("Неверный номер задачи!")
        except ValueError:
            print("Введите число!")

    def show_tasks(self, show_all: bool = True) -> None:
        if not self.tasks:
            print("Список задач пуст!")
            return

        tasks_to_show = self.tasks if show_all else [t for t in self.tasks if not t.completed]
        if not tasks_to_show:
            print("Нет задач для отображения!")
            return

        print("\n=== Список задач ===")
        for i, task in enumerate(tasks_to_show, 1):
            print(f"{i}. {task}")
            if task.description:
                print(f"   Описание: {task.description}")
            if task.completed_at:
                print(f"   Выполнено: {task.completed_at}")

    def show_statistics(self) -> None:
        if not self.tasks:
            print("Список задач пуст!")
            return

        total = len(self.tasks)
        completed = sum(1 for t in self.tasks if t.completed)
        pending = total - completed
        
        print("\n=== Статистика ===")
        print(f"Всего задач: {total}")
        print(f"Выполнено: {completed}")
        print(f"Осталось: {pending}")
        
        if total > 0:
            completion_rate = (completed / total) * 100
            print(f"Процент выполнения: {completion_rate:.1f}%")
        
        if pending > 0:
            print("\nЗадачи с высоким приоритетом:")
            high_priority = [t for t in self.tasks if t.priority >= 4 and not t.completed]
            for task in high_priority:
                print(f"- {task}")

    def clear_screen(self) -> None:
        os.system('cls' if os.name == 'nt' else 'clear')

    def show_menu(self) -> None:
        print("\n=== Менеджер задач ===")
        print("1. Добавить задачу")
        print("2. Редактировать задачу")
        print("3. Удалить задачу")
        print("4. Изменить статус задачи")
        print("5. Показать все задачи")
        print("6. Показать активные задачи")
        print("7. Показать статистику")
        print("0. Выход")
        print("Выберите действие: ", end="")

    def run(self) -> None:
        while True:
            self.clear_screen()
            self.show_menu()
            
            try:
                choice = int(input())
                
                if choice == 0:
                    print("До свидания!")
                    break
                elif choice == 1:
                    self.add_task()
                elif choice == 2:
                    self.edit_task()
                elif choice == 3:
                    self.delete_task()
                elif choice == 4:
                    self.toggle_task()
                elif choice == 5:
                    self.show_tasks(show_all=True)
                elif choice == 6:
                    self.show_tasks(show_all=False)
                elif choice == 7:
                    self.show_statistics()
                else:
                    print("Неверный выбор!")
                
                input("\nНажмите Enter для продолжения...")
            except ValueError:
                print("Пожалуйста, введите число!")
                input("\nНажмите Enter для продолжения...")

if __name__ == "__main__":
    manager = TaskManager()
    manager.run() 